### ISSUE TITLE
Parser service is using 'tsconfig.json'

### ISSUE THREAD

**Issue by armano2** [JamesHenry/typescript-estree#50]
**What version of TypeScript are you using?**
~3.2.1

**What version of `typescript-estree` are you using?**
5.3.0 and tested on master

**What code were you trying to parse?**

```ts
var foo = parseInt("5.5", 10) + 10;
```
```json
{
    "compilerOptions": {
        "target": "es5",
        "module": "commonjs",
        "strict": true,
        "esModuleInterop": true,
        "lib": [
            "es2015.core"
        ]
    }
}
```

**What did you expect to happen?**
typeChecker has no informations about types from `es2015.core`

**Demo**
https://github.com/bradzacher/eslint-plugin-typescript/pull/230

**Related to**
https://github.com/JamesHenry/typescript-estree/pull/22

**Comment by armano2** [comment]
Update: issue happens only when filePath is not specified

**Comment by uniqueiniquity** [comment]
@armano2 I'll take a look

**Comment by armano2** [comment]
- typescript requires that file is provided if you are using tsconfig
- typescript requires that file is in matching tsconfig configuration (resolving)

solution:
- allow to pass parser configuration as object

**Comment by uniqueiniquity** [comment]
What do you mean by "passing parser configuration"?
The way the TypeScript compiler test harness handles this is by providing a virtual file system and copy of the library. Is there a more limited approach that you have in mind?

**Comment by armano2** [comment]
```ts
// create compiler host
    const watchCompilerHost = ts.createWatchCompilerHost(
      tsconfigPath,
      /*optionsToExtend*/ undefined,
      ts.sys,
      ts.createSemanticDiagnosticsBuilderProgram,
      diagnosticReporter,
      /*reportWatchStatus*/ () => {}
    );
```

`/*optionsToExtend*/ undefined,`

**Comment by armano2** [comment]
I have issues with making test for eslint in RuleTester, because i'm unable to provide configuration file (i'm not making files there, it's just code)

**Comment by uniqueiniquity** [comment]
It appears to me that even with the ability to pass TypeScript compiler options there, the problem still occurs. In particular, when passing an object with the `lib` option set, it doesn't look for the correct paths. Were you able to get it to work this way?

Instead of giving `typescript-estree` the ability to take a config object, I'd prefer that we try to set up the `eslint-plugin-typescript` test infrastructure to better support this. We could either do that with your approach, with an `empty.ts` file and a `tsconfig.json` that includes it, or with a more complicated virtual file system.

**Comment by armano2** [comment]
i think issue is somewhere else, we should execute `createCompilerHost` instead of `createWatchCompilerHost` when instead of file we are getting sourceCode:
```
function createCompilerHost(options: CompilerOptions, setParentNodes?: boolean): CompilerHost;
```

**Comment by uniqueiniquity** [comment]
How would that help? We do actually need a `WatchCompilerHost` in order to efficiently handle multiple parse operations over a single process lifetime.

**Comment by armano2** [comment]
`WatchCompilerHost` is working correctly when there is `file` provided. And this is correct when we want to watch for file changes.

**Comment by uniqueiniquity** [comment]
I'm not sure I understand you completely. Let me try to summarize what I think you are saying; please tell me if I am not correct.

MOTIVATION:
In `eslint-plugin-typescript`, we want to test semantic rules. However, test cases that use functions from `lib` don't work as easily as existing test cases. This is because `typescript-estree` requires an actual `tsconfig.json` file to specify compiler options, and for those options to apply, the test file must exist and be included by the `tsconfig.json`. 

PROPOSAL:
Your suggestion is to change the `typescript-estree` API so that, as an alternative to reading a particular `tsconfig.json` file, users of this package can provide a TypeScript configuration options object to the `parse` method.

CONCERNS:
I think this will only solve the issue of needing a real `tsconfig.json` object for the test runner in `eslint-plugin-typescript`. As far as I have seen, the `lib` files do not resolve properly when they are provided through `optionsToExtend` in `createWatchCompilerhost`.

Regarding your most recent comment, `shouldProvideParserServices` is not a recognized option for `typescript-estree`. It is only an internal flag for distinguishing between a call to `generateAST` through `parse` versus `parseAndGenerateServices`. The tests in https://github.com/JamesHenry/typescript-estree/blob/master/tests/lib/semanticInfo.ts do use `parseAndGenerateServices`, and therefore do have `shouldGenerateServices` and `project` set (and as a result, have `shouldProvideParserServices` set as well).

I am certainly open to reviewing changes to `typescript-estree` to address this issue, though I think @JamesHenry should weigh in on any API changes. However, since this issue seems specific to the `eslint-plugin-typescript` test infrastructure, I think we should try to resolve it in that package.

**Comment by armano2** [comment]
@uniqueiniquity i did some testings with setup with 
[`vue-eslint-parser`](https://github.com/mysticatea/vue-eslint-parser) +  `eslint-plugin-typescript` and looks like this issue is present

**Comment by uniqueiniquity** [comment]
Yep, my understanding is that #53 addressed the issues discussed here.
